### PR TITLE
Add support for categorised codelists

### PR DESCRIPTION
**Pull request by rebkwok**

### CODE REVIEW

**Review by benbc — COMMENTED** [review]
I'm uncomfortable with this -- it seems to be very invasive. I'm very tired today due to my son having had too much fun camping last night and I'm not sure that I've really grokked it well enough to propose alternatives, but I do have a few thoughts.

It would be helpful to have a concrete example of how these code categories are used in real cohorts. Have you seen such a thing?

I think I need to invoke @evansd here, who may already have thought about some of this stuff.

Could we treat codelists as always categorized (but with a null category as a possibility)? This might simplify some of the code where we currently behave different with categorized and uncategorized codelists.

Closely related, what about if we always propagated a "category" column up the chain if we have joined to a codelist table. I have a feeling that would allow us to remove some conditionality.

Rather more radically, this feels a bit like something that I've pondered a bit before. We might express some of what we need to do as a series of global transforms on the DAG. For example breaking off a linear branch of the graph and turning it into a temp table creation, with the "stump" as a select from that table. If we could consider the "invasive" stuff you've had to here as single operation across the entire DAG, it would localize all the related stuff in one place and make it easier to understand.

File: cohortextractor/codelistlib.py [hunk 1]
```diff
@@ -22,42 +36,41 @@ def codelist_from_csv(filename, system, column="code", category_column=None):
...
                 )
             if column not in row:
                 raise ValueError(
                     f"Codelist csv file at {filename} does not contain column '{column}'"
                 )
-            codes.append(row[column].strip())
+            if category_column:
+                codes.append((row[column].strip(), row[category_column].strip()))
+            else:
+                codes.append(row[column].strip())
```
[the remark this answers is not in the document]
Comment (evansd) [reply cohortextractor/codelistlib.py:50]: This looks like it comes directly from the old cohortextractor which had the simplest possible implementation of codelists and then had various features (like categories) rapidly hacked on to it with the intention of "doing it properly soon". But as it worked OK and never needed any more features that day never came.

I think the simplest improvement (which you mentioned elsewhere) would be to treat all codelists as categorised and just assign null categories if we don't have them.

**Comment by evansd** [comment]
This is an extremely tricky feature to implement, to the extent that I don't even know what the ideal DSL would look like for it yet. There's something quite unpleasant about the idea that doing a certain sort of `filter()` operation magically generates a new column that can then be used with `get()`. I've wondered about having an explicit method which adds the column

```py
  ethnicity = table("coded_events")
     .add_category_column("ethnicity", from="code", using=ethnicity_codelist)
    .latest()
    .get("ethnicity")
```

I'd really like to be do without it altogether but it's heavily used at the moment:
https://github.com/search?q=org%3Aopensafely+returning%3D%22category%22&type=code

A lot of those uses could be replaced by extracting the code and then doing the category mapping after the data has been downloaded as a post-processing step. But I think sometimes the categories are used as an input to other functions which we wouldn't be able to do using that method.

It might be worth brainstorming a bit with Alex as to what other alternatives there might be.

File: cohortextractor/codelistlib.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rebkwok) [reply cohortextractor/codelistlib.py:50]: It does come directly from the old codelists.  

Treating all codelists as categorised is fine, but there would still be the issue of whether a `foo.get("category")` could conflict with another column in the original table named `category`.  We do control the backend tables though, so we would know if that was a possibility.  Maybe just naming it `codelist_category` is enough?

**Comment by rebkwok** [comment]
I tried to port over what we had in the old cohort-extractor as much as possible, but it sounds like this is one bit that we don't want to do that for.  Happy to try and address the issues, but maybe this needs to wait on more discussion - ethnicity is the only one that needs categorisation for the real study, and it's not used as the input to anything else.

**Comment by benbc** [comment]
We're going to reconsider the design here when we actually need it.

**Pull request closed without merging**